File: packages/kotlin-sdk/KotlinExampleApp/app/src/main/java/org/dashfoundation/example/services/tokens/TokenDirectPurchasePricing.kt
package org.dashfoundation.example.services.tokens

import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import org.dashfoundation.example.util.LenientJson
import java.math.BigInteger

/**
 * A token's direct-purchase pricing schedule, parsed from the JSON that
 * `TokenQueries.directPurchasePrices` returns (shape emitted by the FFI's
 * `dash_sdk_token_get_direct_purchase_prices`):
 *
 * ```
 * {"<tokenIdBase58>": {"type":"single_price","price":<u64>}}
 * {"<tokenIdBase58>": {"type":"set_prices","prices":[{"amount":<u64>,"price":<u64>}, …]}}
 * {"<tokenIdBase58>": null}
 * ```
 *
 * This is the model behind the Direct Purchase form's cost preview. It
 * resolves the total credits a buyer must pay for a given raw token amount
 * using the *same* tier rule Drive applies when it validates the purchase
 * (`token_direct_purchase_transition_action` v0 transformer), so the
 * `expectedTotalCost` the form submits matches the chain's `required_price`:
 *
 *  - [SinglePrice]: `price × amount`.
 *  - [SetPrices]: the highest tier threshold `≤ amount` sets the per-token
 *    price and the total is `tierPrice × amount`. An amount below the
 *    smallest threshold is under the minimum sale amount (not purchasable);
 *    an empty schedule means the token isn't for direct sale.
 *
 * Prices are u64 credits on-chain, so parsing and multiplication go through
 * [BigInteger]. Drive saturates an overflowing [SinglePrice] multiplication,
 * while a [SetPrices] multiplication uses checked arithmetic and rejects the
 * purchase; this preview mirrors that distinction. A zero price is valid and
 * produces a zero expected cost, matching the native purchase contract.
 */
sealed interface TokenDirectPurchasePricing {

    /** One flat per-token price for any amount. */
    data class SinglePrice(val price: BigInteger) : TokenDirectPurchasePricing

    /** Tiered pricing: threshold → per-token price. */
    data class SetPrices(val tiers: List<Tier>) : TokenDirectPurchasePricing {
        /** A `(minimum amount, per-token price)` tier. */
        data class Tier(val amount: BigInteger, val price: BigInteger)
    }

    /**
     * Required total cost in credits to buy [amount] raw token units, or
     * `null` when the amount isn't purchasable at this schedule (below the
     * minimum sale amount, an empty schedule, or a tiered-price overflow.
     * Single-price overflow saturates to u64 max, matching Drive.
     */
    fun costFor(amount: ULong): ULong? {
        if (amount == 0uL) return null
        val count = BigInteger(amount.toString())
        if (count > MAX_TOKEN_COUNT) return null
        val (perToken, saturatesOnOverflow) = when (this) {
            is SinglePrice -> price to true
            is SetPrices -> (
                tiers
                    .filter { it.amount <= count }
                    .maxByOrNull { it.amount }
                    ?.price ?: return null
                ) to false
        }
        val total = perToken * count
        if (total.signum() < 0) return null
        if (total > MAX_U64 && !saturatesOnOverflow) return null
        return total.min(MAX_U64).toString().toULong()
    }

    /**
     * The smallest raw amount that can be purchased — `1` for a single
     * price, the lowest tier threshold for a tiered schedule. Used for the
     * "minimum purchase is N" hint.
     */
    val minimumPurchaseAmount: BigInteger
        get() = when (this) {
            is SinglePrice -> BigInteger.ONE
            is SetPrices -> tiers.minOfOrNull { it.amount } ?: BigInteger.ONE
        }

    companion object {
        private val MAX_U64 = BigInteger(ULong.MAX_VALUE.toString())
        // DPP's MAX_DISTRIBUTION_PARAM: token transition amounts are u48,
        // even though pricing and balance carriers use the full u64 domain.
        private val MAX_TOKEN_COUNT = BigInteger("281474976710655")

        /**
         * Parse the pricing for [canonicalTokenId] out of a
         * `directPurchasePrices` response, or `null` when the token has no
         * usable price (a `null`/missing entry, an empty tier list) or the
         * JSON can't be read.
         */
        fun parse(json: String, canonicalTokenId: String): TokenDirectPurchasePricing? {
            val entry = runCatching {
                LenientJson.parseToJsonElement(json).jsonObject[canonicalTokenId]
            }.getOrNull()
            val obj = entry as? JsonObject ?: return null
            return runCatching {
                when (obj["type"]?.jsonPrimitive?.content) {
                    "single_price" -> obj["price"]?.bigIntegerOrNull()?.let { SinglePrice(it) }
                    "set_prices" -> {
                        val tiers = obj["prices"]?.jsonArray.orEmpty().mapNotNull { element ->
                            val tier = element as? JsonObject ?: return@mapNotNull null
                            val amount = tier["amount"]?.bigIntegerOrNull()
                            val price = tier["price"]?.bigIntegerOrNull()
                            if (amount != null && price != null) {
                                SetPrices.Tier(amount, price)
                            } else {
                                null
                            }
                        }
                        if (tiers.isEmpty()) null else SetPrices(tiers)
                    }
                    else -> null
                }
            }.getOrNull()
        }

        private fun JsonElement.bigIntegerOrNull(): BigInteger? =
            runCatching { BigInteger(jsonPrimitive.content) }.getOrNull()
    }
}

File: packages/swift-sdk/SwiftExampleApp/SwiftExampleApp/Utils/TokenDirectPurchasePricing.swift
import Foundation

/// A token's direct-purchase pricing schedule, parsed from the JSON that
/// `SDK.getTokenDirectPurchasePrices(tokenIds:)` returns (the shape emitted
/// by the FFI's `dash_sdk_token_get_direct_purchase_prices`):
///
/// ```
/// {"<tokenIdBase58>": {"type":"single_price","price":<u64>}}
/// {"<tokenIdBase58>": {"type":"set_prices","prices":[{"amount":<u64>,"price":<u64>}, …]}}
/// {"<tokenIdBase58>": null}
/// ```
///
/// This is the model behind the Direct Purchase form's cost preview. It
/// resolves the total credits a buyer must pay for a given raw token amount
/// using the *same* tier rule Drive applies when it validates the purchase
/// (`token_direct_purchase_transition_action` v0 transformer), so the
/// `expectedTotalCost` the form submits matches the chain's `required_price`:
///
///  - ``singlePrice(_:)``: `price × amount`.
///  - ``setPrices(_:)``: the highest tier threshold `≤ amount` sets the
///    per-token price and the total is `tierPrice × amount`. An amount below
///    the smallest threshold is under the minimum sale amount (not
///    purchasable); an empty schedule means the token isn't for direct sale.
///
/// Prices are `u64` credits on-chain, so multiplication goes through
/// `multipliedReportingOverflow`; a product that overflows `UInt64` resolves
/// to `nil` (Buy stays disabled) rather than wrapping. A resolved cost of `0`
/// (a free tier) is likewise `nil`: the purchase FFI/flow requires a positive
/// `expectedTotalCost`.
enum TokenDirectPurchasePricing: Equatable {

    /// One flat per-token price for any amount.
    case singlePrice(UInt64)

    /// Tiered pricing: an ordered list of `(minimum amount, per-token price)`
    /// tiers. The highest tier whose `amount ≤` the purchased amount sets the
    /// per-token price.
    case setPrices([Tier])

    /// A `(minimum amount, per-token price)` tier in a ``setPrices(_:)``
    /// schedule.
    struct Tier: Equatable {
        let amount: UInt64
        let price: UInt64
    }

    /// The required total cost in credits to buy `amount` raw token units, or
    /// `nil` when the amount isn't purchasable at this schedule (below the
    /// minimum sale amount, a free tier, or a total that overflows the
    /// `UInt64` credits the purchase FFI accepts).
    ///
    /// This mirrors Drive's `token_direct_purchase_transition_action` v0
    /// transformer: `required_price = perTokenPrice × token_count`, where for
    /// a tiered schedule the per-token price is the highest tier threshold
    /// `≤ amount`.
    ///
    /// One deliberate divergence: Drive's single-price branch uses
    /// `saturating_mul`, so an overflowing product passes its price check at
    /// `u64::MAX` — but Drive then debits that `u64::MAX`, which exceeds any
    /// possible credit balance, so the transition is guaranteed to fail at
    /// balance deduction (after charging a processing fee). Broadcasting a
    /// known-doomed purchase is worse than disabling Buy, so this helper
    /// treats overflow as not-purchasable in both branches (as the Kotlin
    /// counterpart does — its `Long`-based FFI can't even represent
    /// `u64::MAX`).
    func cost(forAmount amount: UInt64) -> UInt64? {
        guard amount > 0 else { return nil }

        let perToken: UInt64
        switch self {
        case let .singlePrice(price):
            perToken = price
        case let .setPrices(tiers):
            // Highest tier threshold `≤ amount` — the Rust
            // `set_prices.range(..=token_count).next_back()` lookup. An amount
            // below every threshold (or an empty schedule) is not purchasable.
            guard let tierPrice = tiers
                .filter({ $0.amount <= amount })
                .max(by: { $0.amount < $1.amount })?
                .price
            else { return nil }
            perToken = tierPrice
        }

        let (total, overflow) = perToken.multipliedReportingOverflow(by: amount)
        guard !overflow else { return nil }
        // A resolved total of 0 (a free tier) is not submittable: the purchase
        // FFI/flow requires a positive expected total cost.
        guard total > 0 else { return nil }
        return total
    }

    /// The smallest raw amount that can be purchased — `1` for a single price,
    /// the lowest tier threshold for a tiered schedule. Used for the "minimum
    /// purchase is N" hint.
    var minimumPurchaseAmount: UInt64 {
        switch self {
        case .singlePrice:
            return 1
        case let .setPrices(tiers):
            return tiers.map { $0.amount }.min() ?? 1
        }
    }

    /// Parse the pricing for `canonicalTokenId` out of the dictionary a
    /// `getTokenDirectPurchasePrices` call produces (a `[String: Any]` from
    /// `JSONSerialization`), or `nil` when the token has no usable price (a
    /// `null`/missing entry, an empty tier list) or the entry can't be read.
    ///
    /// `price`/`amount` values are `u64` and can exceed `Int64.max`, so they
    /// are read through `NSNumber.uint64Value` rather than as `Int`.
    static func parse(
        _ response: [String: Any],
        canonicalTokenId: String
    ) -> TokenDirectPurchasePricing? {
        // A missing key or an explicit JSON `null` (bridged to `NSNull`) both
        // mean "no price configured".
        guard let entry = response[canonicalTokenId] as? [String: Any] else {
            return nil
        }

        switch entry["type"] as? String {
        case "single_price":
            guard let price = (entry["price"] as? NSNumber)?.uint64Value else {
                return nil
            }
            return .singlePrice(price)

        case "set_prices":
            guard let rawTiers = entry["prices"] as? [[String: Any]] else {
                return nil
            }
            let tiers: [Tier] = rawTiers.compactMap { tier in
                guard let amount = (tier["amount"] as? NSNumber)?.uint64Value,
                      let price = (tier["price"] as? NSNumber)?.uint64Value
                else { return nil }
                return Tier(amount: amount, price: price)
            }
            // An empty schedule (or one whose every tier failed to parse) means
            // the token isn't for direct sale.
            return tiers.isEmpty ? nil : .setPrices(tiers)

        default:
            return nil
        }
    }
}
